---
title: "Analysis of singlecellRNAseq data by 10x"
author: "Ivo Borkus"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:  
  pdf_document: default
  html_document:
    toc: true
    toc_depth: 1
---
```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/scratch_isilon/groups/singlecell/iborkus/Projects/brain_met") # nolint
knitr::opts_chunk$set(warning = FALSE)
```
```{R loading libraries, include = FALSE}
library(Seurat)
library(dplyr)
library(ggplot2)
library(readr)
library(purrr)
library(tibble)
library(here)
library(tidyr)
library(paletteer)
library(ggrepel)
library(gridExtra)
setwd("/scratch_isilon/groups/singlecell/iborkus/Projects/brain_met")
here::i_am("02_scripts/06_doublet_identification.rmd")
set.seed(345)
```

```{R Loading data, echo = FALSE}
meta_data <- read.csv(here("03_processing/01_QC/data/meta_data.csv"))
seurat_obj <- readRDS(file = "03_processing/01_QC/data/filtered_merged_seurat.rds")
nejm_colors <- paletteer::paletteer_d("ggsci::default_nejm", length(levels(as.factor(meta_data$sample_name))))
names(nejm_colors) <- levels(meta_data$sample_name)
# checking memory usage
Rprof(tmp <- tempfile(), memory.profiling = TRUE)
## Layers are already split:
# seurat_obj[["RNA"]] <- split(seurat_obj[["RNA"]],f = seurat_obj$sample)
```


## Doublet_finder needs a preprocessed dataframe:

-  Do not apply DoubletFinder to aggregated scRNA-seq data representing multiple distinct samples (e.g., multiple 10X lanes). For example, if you run DoubletFinder on aggregated data representing WT and mutant cell lines sequenced across different 10X lanes, artificial doublets will be generated from WT and mutant cells, which cannot exist in your data. These artificial doublets will skew results. Notably, it is okay to run DoubletFinder on data generated by splitting a single sample across multiple 10X lanes.
-  Ensure that input data is cleared of low-quality cell clusters. There are a variety of ways to do this, but I usually use the following workflow:

1. Manually threshold raw gene expression matrices according to RNA nUMIs especially important when dealing with super-loaded 10X data because of the way CellRanger threholds data -- See Lun et al., 2019, Genome Biology.
2. Pre-process data using standard workflow.
3. Identify clusters with (A) low RNA UMIs, (B) High % mitochondrial reads, and/or (C) Uninformative marker genes.
4. Remove clusters, pre-process again, and run DoubletFinder.

```{R finding variable features, echo = FALSE}
violin_preNorm <- VlnPlot(seurat_obj, features = "MOG") + NoLegend()
ggsave(filename = "03_processing/02_Norm/figs/violin_preNorm_01_unintegrated.pdf", plot = violin_preNorm)
Layers(seurat_obj[["RNA"]])

str(seurat_obj)
seurat_obj <- NormalizeData(
    object = seurat_obj,
    normalization.method = "LogNormalize",
    scale.facor = 10000
)
violin_postNorm <- VlnPlot(seurat_obj, features = "MOG") + NoLegend()
ggsave(filename = "03_processing/02_Norm/figs/violin_postNorm_01_unintegrated.pdf", plot = violin_postNorm)

seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(seurat_obj)
seurat_obj <- ScaleData(object = seurat_obj, features = all.genes)
Variable_genes_plot <- LabelPoints(
    plot = VariableFeaturePlot(seurat_obj),
    points = head(VariableFeatures(seurat_obj), 10),
    repel = FALSE
)


VariableFeatures(seurat_obj)

str(seurat_obj)
ggsave(filename = "03_processing/03_Scale/figs/Variable_genes_plot_unintegrated.pdf")
str(seurat_obj)
```

```{R running PCA, echo=FALSE}
seurat_obj <- RunPCA(object = seurat_obj, features = VariableFeatures(object = seurat_obj), nfeatures.print = 5, ndims.print = 1:2) ## Using the 2000 features instead of all genes
```
```{R Plotting}
Genes_influence_PCA <- VizDimLoadings(seurat_obj, dims = 1:5, reduction = "pca")
PCA_dim_plot <- DimPlot(seurat_obj, reduction = "pca", group.by = "sample") + NoLegend()
PCA_gene_heatmap <- DimHeatmap(seurat_obj, dims = 1:10, cells = 100, balanced = TRUE) #
PCA_elbow <- ElbowPlot(seurat_obj)

Genes_influence_PCA ## Top genes contributing to the PC
PCA_dim_plot ## PCA projection of data
PCA_gene_heatmap ## Heatmap showing top genes that are influencing each pc and the highest and lowest scoring cells.
PCA_elbow ## Showing the amount of variability that is given by each PC
ggsave(filename = "03_processing/04_PCA/figs/Genes_influence_PCA.pdf", plot = Genes_influence_PCA)
ggsave(filename = "03_processing/04_PCA/figs/PCA_dim_plot.pdf", plot = PCA_dim_plot)
ggsave(filename = "03_processing/04_PCA/figs/PCA_gene_heatmap.pdf", plot = PCA_gene_heatmap)
ggsave(filename = "03_processing/04_PCA/figs/PCA_elbow.pdf", plot = PCA_elbow)
```

```{R}
seurat_obj <- FindNeighbors(seurat_obj, dims = 1:10, reduction = "pca")
seurat_obj <- FindClusters(seurat_obj, resolution = 0.5, algorithm = 4, cluster.name = "unintegrated_clusters")
seurat_obj <- RunUMAP(seurat_obj, dims = 1:10, reduction = "pca", reduction.name = "umap.unintegrated")
```

```{R QC of clusters}
umap <- DimPlot(
    object = seurat_obj,
    label = TRUE, reduction = "umap.unintegrated", split.by = "sample", pt.size = 0.05, cols = nejm_colors
) &
    theme(plot.title = element_text(size = 10)) & NoAxes()
# "seurat_clusters"
```
```{R printing memore usage,echo=TRUE}
Rprof(NULL)
cat(head(summaryRprof(tmp, memory = "both", lines = "hide")$by.total))
```